src/upsampler.rs

```diff
@@ -34,12 +35,14 @@ impl Upsampler {
 
         Ok(Upsampler {
             components: upsampler_components,
+            max_horizontal_sampling_factor: h_max
         })
     }
 
     pub fn upsample_and_interleave_row(&self, component_data: &[Vec<u8>], row: usize, output_width: usize, output: &mut [u8]) {
         let component_count = component_data.len();
-        let mut line_buffer = vec![0u8; output_width + 1];
+        let max_component_width = self.components.iter().map(|c| c.width).max().unwrap();
+        let mut line_buffer = vec![0u8; max_component_width * self.max_horizontal_sampling_factor as usize];
```

Comment:
Here I would like to cache the result of `components.iter().map(|c| c.size.width as usize * c.horizontal_sampling_factor as usize).max().unwrap()` in Upsampler and use that for line_buffer's size instead.